src/libraries/Microsoft.Bcl.Cryptography/src/System/Security/Cryptography/PemEncoding.cs

```diff
@@ -270,6 +271,145 @@ private static bool IsWhiteSpaceCharacter(char ch)
+        {
+            // The largest possible label is MaxLabelSize - when included in the posteb
+            // and preeb lines new lines, assuming the base64 content is empty.
+            //     -----BEGIN {char * MaxLabelSize}-----\n
+            //     -----END {char * MaxLabelSize}-----
+            const int MaxLabelSize = 1_073_741_808;
+
+            // The largest possible binary value to fit in a padded base64 string
+            // is 1,610,612,733 bytes. RFC 7468 states:
+            //   Generators MUST wrap the base64-encoded lines so that each line
+            //   consists of exactly 64 characters except for the final line
+            // We need to account for new line characters, every 64 characters.
+            // This works out to 1,585,834,053 maximum bytes in data when wrapping
+            // is accounted for assuming an empty label.
+            const int MaxDataLength = 1_585_834_053;
+
+            if (labelLength < 0)
+                throw new ArgumentOutOfRangeException(nameof(labelLength), SR.ArgumentOutOfRange_NeedPositiveNumber);
+            if (dataLength < 0)
+                throw new ArgumentOutOfRangeException(nameof(dataLength), SR.ArgumentOutOfRange_NeedPositiveNumber);
+            if (labelLength > MaxLabelSize)
+                throw new ArgumentOutOfRangeException(nameof(labelLength), SR.Argument_PemEncoding_EncodedSizeTooLarge);
+            if (dataLength > MaxDataLength)
+                throw new ArgumentOutOfRangeException(nameof(dataLength), SR.Argument_PemEncoding_EncodedSizeTooLarge);
+
+            int preebLength = PreEBPrefix.Length + labelLength + Ending.Length;
+            int postebLength = PostEBPrefix.Length + labelLength + Ending.Length;
+            int totalEncapLength = preebLength + postebLength + 1; //Add one for newline after preeb
+
+            // dataLength is already known to not overflow here
+            int encodedDataLength = ((dataLength + 2) / 3) << 2;
+            int lineCount = Math.DivRem(encodedDataLength, EncodedLineLength, out int remainder);
+
+            if (remainder > 0)
+                lineCount++;
+
+            int encodedDataLengthWithBreaks = encodedDataLength + lineCount;
+
+            if (int.MaxValue - encodedDataLengthWithBreaks < totalEncapLength)
+                throw new ArgumentException(SR.Argument_PemEncoding_EncodedSizeTooLarge);
+
+            return encodedDataLengthWithBreaks + totalEncapLength;
+        }
+
+        internal static bool TryWrite(ReadOnlySpan<char> label, ReadOnlySpan<byte> data, Span<char> destination, out int charsWritten)
+        {
+            static int Write(ReadOnlySpan<char> str, Span<char> dest, int offset)
+            {
+                str.CopyTo(dest.Slice(offset));
+                return str.Length;
+            }
+
+            static int WriteBase64(ReadOnlySpan<byte> bytes, Span<char> dest, int offset, byte[] bytesBuffer)
```

Comment:
This is one place where I made a slight adjustment from the .NET 5 implementation. This version of it must used `Convert.ToBase64String`, not the span-writing one. However, the array one _can_ work with an index and offset. So we can use a rented buffer.

The `WriteBase64` helper now needs `bytesBuffer`, which it will use as a scratch buffer to copy span data to.